src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -227,8 +361,502 @@ sealed abstract class AST(pos: Position, subexprs: Array[AST] = Array.empty) {
+    var mresult = CM.ret(Array[T]() : IndexedSeq[T])
+    mts.foreach(mt => mresult = mresult.flatMap(r => mt.map(t => t +: r)))
+    mresult.map(_.reverse)
+  }
+  def ret[T](t: T): CM[T] = CM((e,a) => (t, a))
+
+  def fb(): CM[Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef]] = CM { (e, s) => (e.fb, s) }
+  def availableBindings(): CM[Map[String, Code[AnyRef]]] = CM { case (e, s) => (e.m, s) }
+
+  def addForeignFun(f: AnyRef): CM[Int] = CM { (e, s) => (s.fa.length, s.copy(fa = f +: s.fa)) }
+  def foreignFunArray(): CM[Code[Array[AnyRef]]] = fb().map(_.arg1)
+  def foreignFun[T <: AnyRef](x: T): CM[Code[T]] = for (
+    i <- addForeignFun(x);
+    ffa <- foreignFunArray()
+  ) yield ffa(i).asInstanceOf[Code[T]]
+
+  def initialValueArray(): CM[Code[mutable.ArrayBuffer[AnyRef]]] = fb().map(_.arg2)
+
+  def ec(): CM[EvalContext] = CM { (e, s) => (s.ec, s) }
+  def currentSymbolTable(): CM[SymbolTable] = CM { (e, s) => (s.ec.st, s) }
+  def ecNewPosition(): CM[(Int, mutable.ArrayBuffer[Any])] = CM { (e, s) =>
+    val idx = s.ec.a.length
+    val localA = s.ec.a
+    localA += null
+    ((idx, localA), s)
+  }
+  // returns a thunk that looks up the result of this aggregation
+  def addAggregation(lhs: AST, agg: Aggregator): CM[() => Any] = CM { (e, s) =>
+    val b = RefBox(null)
+    s.ec.aggregations += ((b, lhs.runAggregator(s.ec), agg))
+    (() => b.v, s)
+  }
+
+  def newLocal[T](implicit tti: TypeInfo[T]): CM[LocalRef[T]] = fb().map(_.newLocal[T])
+  def memoize[T](mc: CM[Code[T]])(implicit tti: TypeInfo[T]): CM[(Code[Unit], Code[T])] = for (
+    c <- mc;
+    x <- newLocal[T]
+  ) yield (x.store(c), x.load().asInstanceOf[Code[T]])
+  def memoize[T](c: Code[T])(implicit tti: TypeInfo[T]): CM[(Code[Unit], Code[T])] = for (
+    x <- newLocal[T]
+  ) yield (x.store(c), x.load().asInstanceOf[Code[T]])
+
+  // references to its argument will be duplicated
+  def bindInRaw[T](name: String, c: Code[AnyRef])(body: CM[Code[T]]): CM[Code[T]] = CM { case (e, s) =>
+    body.mt(e.copy(m = e.m + ((name, c))), s)
+  }
+  def bindRepInRaw[T](bindings: Seq[(String, CM[Code[AnyRef]])])(body: CM[Code[T]]): CM[Code[T]] = bindings match {
+    case Seq() => body
+    case Seq((name, cmc), rest @ _*) =>
+      cmc.flatMap(c => bindInRaw(name, c)(bindRepInRaw(rest)(body)))
+  }
+  // its argument is stored in a local variable, only refernece to the variable
+  // are dupliacted
+  def bindIn[T](name: String, c: Code[AnyRef])(body: CM[Code[T]]): CM[Code[T]] =
+    memoize(c).flatMap { case (st, v) => bindInRaw(name, v)(body.map(x => Code(st, x))) }
+  def bindRepIn[T](bindings: Seq[(String, CM[Code[AnyRef]])])(body: CM[Code[T]]): CM[Code[T]] = bindings match {
+    case Seq() => body
+    case Seq((name, cmc), rest @ _*) =>
+      cmc.flatMap(c => bindIn(name, c)(bindRepIn(rest)(body)))
+  }
+  def lookup(name: String): CM[Code[AnyRef]] = CM { case (e, s) =>
+    (e.m(name), s)
+  }
+
+  def invokePrimitive1Pure[A,R](f: (A) => R)
```

Comment:
Probably don't need this anymore? All the ones below should really have this type since it's weaker, but it seems to be harder to use actually!